### PR TITLE
feat(query, coordinator): move offset & lookback to ExecPlan and some bug fixes

### PR DESCRIPTION
**Pull request by TanviBhavsar**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

- Move offset logic from LogicalPlan to ExecPlan. 
- Move lookback logic from LogicalPlan to QueryPlanner
- Fix FiloCli by passing empty config instead of routing config.

### CODE REVIEW

File: prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala [hunk 1]
```diff
@@ -128,7 +128,7 @@ trait Functions extends Base with Operators with Vectors {
           // Special FiloDB functions to extract things like chunk metadata
         } else if (filoFunctionIdOpt.isDefined) {
           // No lookback needed as we are looking at chunk metadata only, not raw samples
-          val rangeSelector = timeParamToSelector(timeParams, 0)
+          val rangeSelector = timeParamToSelector(timeParams)
```
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala:131]: Remove "no lookback needed" comment. Outdated now.

File: prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala [hunk 2]
```diff
@@ -174,18 +174,15 @@ trait Functions extends Base with Operators with Vectors {
         val rangeFunctionId = RangeFunctionId.withNameInsensitiveOption(name).get
         if (rangeFunctionId == Timestamp) {
           val instantExpression = seriesParam.asInstanceOf[InstantExpression]
-          val offsetMillis: Long = instantExpression.offset.map(_.millis).getOrElse(0)
 
           PeriodicSeriesWithWindowing(instantExpression.toRawSeriesPlan(timeParams),
-            timeParams.start * 1000 - offsetMillis, timeParams.step * 1000, timeParams.end * 1000 - offsetMillis, 0,
+            timeParams.start * 1000, timeParams.step * 1000, timeParams.end * 1000, 0,
```
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala:179]: Rename start to startSecs etc...

File: prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala [hunk 3]
```diff
@@ -14,7 +14,9 @@ object Vectors {
 
 
 object WindowConstants {
-  val staleDataLookbackSeconds = 5 * 60 // 5 minutes
+  val conf = GlobalConfig.defaultsFromUrl
+  val staleDataLookbackMillis = conf.getConfig("filodb.query").
+                                  getDuration("stale-sample-after").toMillis
```
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala:19]: 👍 Thanks for removing the hard-code

File: prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala [hunk 4]
```diff
@@ -14,7 +14,9 @@ object Vectors {
 
 
 object WindowConstants {
-  val staleDataLookbackSeconds = 5 * 60 // 5 minutes
+  val conf = GlobalConfig.defaultsFromUrl
```
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala:17]: use `systemConfig` not `defaultsFromUrl`

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/LogicalPlanUtils.scala [hunk 5]
```diff
@@ -107,4 +108,19 @@ object LogicalPlanUtils {
...
+  def getOffsetMillis(logicalPlan: LogicalPlan): Long = {
+    LogicalPlan.findLeafLogicalPlans(logicalPlan).head match {
+      case lp: RawSeries => lp.offsetMs.getOrElse(0)
+      case _             => 0
+    }
+  }
+
+  def getLookBackMillis(logicalPlan: LogicalPlan): Long = {
+    val staleDataLookbackMillis = conf.getConfig("filodb.query").
+      getDuration("stale-sample-after").toMillis
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/LogicalPlanUtils.scala:120]: Use WindowConstants - no need for config here.

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/LogicalPlanUtils.scala [hunk 6]
```diff
@@ -1,8 +1,10 @@
 package filodb.coordinator.queryplanner
 
+import filodb.core.GlobalConfig
 import filodb.query._
 
 object LogicalPlanUtils {
+  private val conf = GlobalConfig.defaultsFromUrl
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/LogicalPlanUtils.scala:7]: Remove per next comment. Should not be defaultsFromUrl even if it needs to stay

File: prometheus/src/main/scala/filodb/prometheus/query/PrometheusModel.scala [hunk 7]
```diff
@@ -40,7 +43,7 @@ object PrometheusModel {
         }
         ColumnFilter(m.getName, filter)
       }
-      RawSeries(interval, filters, Nil)
+      RawSeries(interval, filters, Nil, Some(queryConfig.getDuration("stale-sample-after").toMillis), None)
```
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/query/PrometheusModel.scala:46]: WindowConstants. Remove conf

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/HighAvailabilityPlanner.scala [hunk 8]
```diff
@@ -66,9 +66,13 @@ class HighAvailabilityPlanner(dsRef: DatasetRef,
...
 
     // lazy because we want to fetch failures only if needed
+    lazy val offsetMillis = LogicalPlanUtils.getOffsetMillis(logicalPlan)
     lazy val periodicSeriesTime = getPeriodicSeriesTimeFromLogicalPlan(logicalPlan)
+    lazy val periodicSeriesTimeWithOffset = TimeRange(periodicSeriesTime.startMs - offsetMillis,
+      periodicSeriesTime.endMs - offsetMillis)
     lazy val lookBackTime = getRawSeriesStartTime(logicalPlan).map(periodicSeriesTime.startMs - _).get
-    lazy val routingTime = TimeRange(periodicSeriesTime.startMs - lookBackTime, periodicSeriesTime.endMs)
+    lazy val routingTime = TimeRange(periodicSeriesTimeWithOffset.startMs - lookBackTime,
+      periodicSeriesTimeWithOffset.endMs - offsetMillis)
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/HighAvailabilityPlanner.scala:75]: Lookback is applied only to startTime. Currently this is used in generating logical plan 
`def timeParamToSelector(timeParam: TimeRangeParams, startOffset: Long, offset: Long = 0): RangeSelector =
    timeParam match {
      case TimeStepParams(start, step, end) => IntervalSelector(start * 1000 - startOffset - offset,
                                               end * 1000 - offset)
      case InMemoryParam(_)                 => InMemoryChunksSelector
      case WriteBuffersParam(_)             => WriteBufferSelector
    }'
Please correct me if I am wrong

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/LongTimeRangePlanner.scala [hunk 9]
```diff
@@ -27,23 +27,26 @@ class LongTimeRangePlanner(rawClusterPlanner: QueryPlanner,
...
           val firstInstantInRaw = lastDownsampleInstant + p.stepMs
 
           val downsampleLp = copyWithUpdatedTimeRange(logicalPlan,
-                                                      TimeRange(p.startMs, lastDownsampleInstant),
+                                                      TimeRange(startWithOffsetMs, lastDownsampleInstant),
                                                       lookbackMs)
           val downsampleEp = downsampleClusterPlanner.materialize(downsampleLp, qContext)
 
-          val rawLp = copyWithUpdatedTimeRange(logicalPlan, TimeRange(firstInstantInRaw, p.endMs), lookbackMs)
+          val rawLp = copyWithUpdatedTimeRange(logicalPlan, TimeRange(firstInstantInRaw, endWithOffsetMs), lookbackMs)
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/LongTimeRangePlanner.scala:49]: Oh yes. Thanks for letting me know! Also changing downsampleLp time and lastDownsampleInstant

File: coordinator/src/test/scala/filodb.coordinator/queryplanner/SingleClusterPlannerSpec.scala [hunk 10]
```diff
@@ -309,5 +310,35 @@ class SingleClusterPlannerSpec extends FunSpec with Matchers with ScalaFutures {
...
+  multiSchemaExec.rangeVectorTransformers(0).isInstanceOf[PeriodicSamplesMapper] shouldEqual(true)
+  val rvt = multiSchemaExec.rangeVectorTransformers(0).asInstanceOf[PeriodicSamplesMapper]
+  rvt.offsetMs.get shouldEqual(300000)
+  rvt.startWithOffset shouldEqual(400000) // (700 - 300) * 1000
+  rvt.endWithOffset shouldEqual (9700000) // (10000 - 300) * 1000
+}
+
+  it("should generate execPlan with offset with window") {
+    val t = TimeStepParams(700, 1000, 10000)
+    val lp = Parser.queryRangeToLogicalPlan("rate(http_requests_total{job = \"app\"}[5m] offset 5m)", t)
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply coordinator/src/test/scala/filodb.coordinator/queryplanner/SingleClusterPlannerSpec.scala:331]: Added assetions

File: prometheus/src/main/scala/filodb/prometheus/query/PrometheusModel.scala (hunk 7 above)
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/query/PrometheusModel.scala:46]: Lets avoid creating config everywhere. Get from one place.

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/HighAvailabilityPlanner.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (vishramachandran) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/HighAvailabilityPlanner.scala:80]: Lets discuss in person.

**Pull request merged**